File: hw06/src/hw06_lib.py
from scanf import scanf

SUFFIXES = {
    't':   12,
    'g':   9,
    'x':   6,
    'meg': 6,
    'k':   3,
    'm':  -3,
    'u':  -6,
    'n':  -9,
    'p':  -12,
    'f':  -15,
    'a':  -18,
}

class Resistor:
    def __init__(self, name, node_p, node_n, res_value):
        self.name = name
        self.node_p = node_p
        self.node_n = node_n
        self.res_value = res_value


def print_resistor(Res):
    print(f"{Res.name}\t{Res.node_p}\t{Res.node_n}\t{Res.res_value:.2e}")

def i_r(Res, v):
    i = v / Res.res_value
    return i

def decode_value(inStr):
    try:
        return float(inStr)
    except:
        pass
    val_tup = scanf("%f%s", inStr)
    if val_tup == None:
        raise ValueError("Invalid input value")

    suffix = val_tup[1].lower()

    if suffix in SUFFIXES.keys():
        mult = SUFFIXES[suffix]
    else:
        raise ValueError("Invalid suffix")

    value = val_tup[0]
    
    value_fin = value * 10 ** mult
    return value_fin

def read_resistor(inStr):
    res_tup = scanf("%s %s %s %s",inStr)
    if res_tup == None:
        raise ValueError("Bad input string")
    elif 'r' not in res_tup[0].lower():
        raise ValueError("Didn't start with R")
    val = decode_value(res_tup[3])
    if val < 0:
        raise ValueError("Negative Resistor Value")
    res = Resistor(res_tup[0], res_tup[1], res_tup[2], val)
    return res

File: hw06/src/hw06_lib.c
// put your definitions here
#include "hw06_lib.h"
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>

void printResistor(struct Resistor *res) {
	printf("%s\t%s\t%s\t%.2e\n", res->name, res->node_p, res->node_n, res->res_val);	
}

float iR(struct Resistor *res, float v) {
	double current = v / res->res_val;
	return current;
}


bool decodeValue(double *value, const char *str) {
	float val;
	char suffixStr[50];
	int mult;
	if (sscanf(str, "%f%s", &val, suffixStr) == 1) {
		suffixStr[0] = 'l';
		suffixStr[1] = '\0';
	}
	char suffStrLow[50];
	char suffixLow;
	for (int i = 0; i < strlen(suffixStr); i++) {
		suffStrLow[i] = tolower(suffixStr[i]);
	}
	if (strlen(suffixStr) > 1) {
		if (suffStrLow[0] == 'm' && suffStrLow[1] == 'e') {
			suffixLow = 'x';
		} else {
			suffixLow = suffStrLow[0];
		}
	} else {
		suffixLow = suffStrLow[0];
	}
	switch (suffixLow) {
		case 't':
			mult = 12;
			break;
		case 'g':
			mult = 9;
			break;
		case 'x':
			mult = 6;
			break;
		case 'k':
			mult = 3;
			break;
		case 'l':
			mult = 0;
			break;
		case 'm':
			mult = -3;
			break;
		case 'u':
			mult = -6;
			break;
		case 'n':
			mult = -9;
			break;
		case 'p':
			mult = -12;
			break;
		case 'f':
			mult = -15;
			break;
		case 'a':
			mult = -18;
			break;
		case '\0':
			mult = 0;
			break;
		default:
			return false;
			break;
	}
	*value = val * pow(10, mult);
	return true;
}


bool readResistor(struct Resistor *res, const char *str) {
	char name_str[10], node_p_str[10], node_n_str[10], val_str[10];
	if (sscanf(str, "%s %s %s %s", name_str, node_p_str, node_n_str, val_str) != 4) {
		return false;
	}
	double value;

	if (name_str[0] != 'R' && name_str[0] != 'r') {
		return false;
	}


	if (decodeValue(&value, val_str) == false) {
//		fprintf(stderr, "Failure to decode resistor string\n");
		return false;
	}
	strcpy(res->name,name_str);
	strcpy(res->node_p,node_p_str);
	strcpy(res->node_n,node_n_str);
	res->res_val = value;

	return true;
}

